src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override bool CanRead
+        {
+            get { return !_disposed; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return !_disposed; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotImplementedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotImplementedException(); }
+            set { throw new NotImplementedException(); }
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                throw new ObjectDisposedException("ProducerConsumerStream");
+            }
+
+            while (count > 0)
+            {
+                if (_currentBuffer == WriteBufferWrapper.EmptyContainer)
+                {
+                    _currentBuffer = await _buffer.Task;
+                }
+
+                int bytesWritten = _currentBuffer.Write(buffer, offset, count);
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                return 0;
+            }
```

Comment:
Why not throw if attempting to read from a disposed content like this?
And if returning a value is acceptable, is zero the right one (vs. -1)?